Add section "Webhook" in "Troubleshooting" (#1309)
kanisterio/kanister: docs/troubleshooting.rst

<https://github.com/kanisterio/kanister/issues>`_. A `mailing list
<https://groups.google.com/forum/#!forum/kanisterio>`_ is also
available if needed.

> Webhook

Reviewer: ```suggestion
Validating Webhook for Blueprints
```

=======

> For the webhook to work, the Kubernetes API needs to connect to port 9443 of the operator. For some clusters, a firewall has to be configured.

Reviewer: ```suggestion
For the Validating Webhook to work, the Kubernetes API Server needs to connect to port ``9443`` of the Kanister operator. If your cluster has firewall setup, it has to be configured to allow that communication.
```

GKE
---
If you get an error, that the webhook can't be reached you need to check if your firewall misses a rule for port 9443::

>   Error from server (InternalError): error when creating "blueprint.yaml": Internal error occurred: failed calling webhook "blueprints.cr.kanister.io": failed to call webhook: Post "https://kanister-kanister-operator.kanister.svc:443/validate/v1alpha1/blueprint?timeout=5s": context deadline exceeded

Reviewer: let's include this in code blocks.

> See https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules and https://github.com/kubernetes/kubernetes/issues/79739

Reviewer: Can you please refer to other files to see how we specify links in the docs.

<https://github.com/kanisterio/kanister/issues>`_. A `mailing list
<https://groups.google.com/forum/#!forum/kanisterio>`_ is also
available if needed.

Validating webhook for Blueprints
=================================

> For the validating webhook to work, the Kubernetes API Server needs to

Reviewer 2: It will be helpful to describe the symptoms of an unreachable webhook. E.g., what is the output of `kubectl apply -f blueprint.yaml`?

connect to port ``9443`` of the Kanister operator. If your cluster has
a firewall setup, it has to be configured to allow that communication.

GKE
---
If you get an error, that the webhook can't be reached you need to check if

> your firewall misses a rule for port 9443:

Reviewer: ```suggestion
your firewall misses a rule for port ``9443``:
```

Reviewer 2: It will be helpful to mention where this error is from i.e. the Kanister operator logs.

Reviewer: @muffl0n I think this is not addressed.

Author: The error (and the command that precedes it) is in the code block below. I feel that this is explanation enough, isn't it?

Reviewer 2: To confirm, is the error part of the kubectl output or the operator logs?

Author: It's the kubectl output. As the kubernetes control plane is not able to reach the webhook, it can't get any output of the operator, yet.

Reviewer: I still think that 9443 should be ``9443`` but its very minor. I think we can get the PR in.

Reviewer 2: Agreed - for consistency sake.

Author: Thanks for the reminder, I missed the part with the backticks. Just added them.

.. code-block:: console

   Error from server (InternalError): error when creating "blueprint.yaml": Internal error occurred: failed calling webhook "blueprints.cr.kanister.io": failed to call webhook: Post "https://kanister-kanister-operator.kanister.svc:443/validate/v1alpha1/blueprint?timeout=5s": context deadline exceeded

> See `GKE: Adding firewall rules for specific use cases <https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules>`_ and `kubernetes/kubernetes: Using non-443 ports for admission webhooks requires firewall rule in GKE <https://github.com/kubernetes/kubernetes/issues/79739>`_

Reviewer: ```suggestion
See `GKE: Adding firewall rules for specific use cases <https://cloud.google.com/kubernetes-engine/docs/how-to/private-clusters#add_firewall_rules>`_ and `kubernetes/kubernetes: Using non-443 ports for admission webhooks requires firewall rule in GKE <https://github.com/kubernetes/kubernetes/issues/79739>`_ for more details.
```

<https://github.com/kanisterio/kanister/issues>`_. A `mailing list
<https://groups.google.com/forum/#!forum/kanisterio>`_ is also
available if needed.

> Validating webhook for Blueprints

Reviewer 3: ```suggestion

Validating webhook for Blueprints
```

=================================
For the validating webhook to work, the Kubernetes API Server needs to
connect to port ``9443`` of the Kanister operator. If your cluster has
a firewall setup, it has to be configured to allow that communication.

GKE
---

> If you get an error while applying a blueprint, that the webhook can't be reached
> you need to check if your firewall misses a rule for port ``9443``:

Reviewer 3: ```suggestion
If you get an error while applying a blueprint, that the webhook can't be reached,
check if your firewall misses a rule for port ``9443``:
```
